File: it-agents/skills/code-architect/references/code-examples/python/template_method.py
"""Template Method — algorithm skeleton in a base class with overridable steps."""

from abc import ABC, abstractmethod


class ReportGenerator(ABC):
    """Defines the fixed algorithm; subclasses fill in the steps."""

    def generate(self, rows: list[str]) -> str:
        parts = [self.header()]
        parts += [self.format_row(row) for row in rows]
        parts.append(self.footer())
        return "\n".join(parts)

    @abstractmethod
    def header(self) -> str:
        ...

    @abstractmethod
    def format_row(self, row: str) -> str:
        ...

    def footer(self) -> str:  # optional hook with a default
        return "(end)"


class MarkdownReport(ReportGenerator):
    def header(self) -> str:
        return "# Report"

    def format_row(self, row: str) -> str:
        return f"- {row}"


class CsvReport(ReportGenerator):
    def header(self) -> str:
        return "value"

    def format_row(self, row: str) -> str:
        return row

    def footer(self) -> str:
        return ""


if __name__ == "__main__":
    print(MarkdownReport().generate(["alpha", "beta"]))
    print("---")
    print(CsvReport().generate(["alpha", "beta"]))

File: it-agents/skills/code-architect/references/code-examples/php/TemplateMethod.php
<?php
// Template Method — algorithm skeleton in a base class with overridable steps.
declare(strict_types=1);

abstract class ReportGenerator
{
    // The template method defines the invariant skeleton.
    final public function generate(array $rows): string
    {
        $out = $this->header();
        foreach ($rows as $row) {
            $out .= $this->formatRow($row);
        }
        $out .= $this->footer();
        return $out;
    }

    abstract protected function formatRow(string $row): string;

    // Steps with sensible defaults (hooks) subclasses may override.
    protected function header(): string { return ''; }
    protected function footer(): string { return ''; }
}

final class CsvReport extends ReportGenerator
{
    protected function header(): string { return "value\n"; }
    protected function formatRow(string $row): string { return $row . "\n"; }
}

final class HtmlReport extends ReportGenerator
{
    protected function header(): string { return "<ul>\n"; }
    protected function formatRow(string $row): string { return "  <li>$row</li>\n"; }
    protected function footer(): string { return "</ul>\n"; }
}

// Demo
$rows = ['alpha', 'beta'];
echo (new CsvReport())->generate($rows);
echo (new HtmlReport())->generate($rows);
